OpenTelemetry (S3M) Host and OS apps (#1878)
SumoLogic/sumologic-documentation: docs/integrations/hosts-operating-systems/opentelemetry/linux-opentelemetry.md

Preconfigured dashboards provide insight into CPU, memory, network, file descriptors, page faults, and TCP connectors. This app uses OpenTelemetry, an open-source collector for the collection of both metrics and log data.

We use the OpenTelemetry collector for Linux metric collection and for collecting Linux system logs. OpenTelemetry collector runs on the Linux machine, and uses the [Host Metric Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver) to obtain Host and process metrics, and the [Sumo Logic OpenTelemetry Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/sumologicexporter) to send the metrics to Sumo Logic. Linux logs are sent to Sumo Logic through a [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver).

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Linux-Schematics.png')} alt="Schematics" />

## Fields Created in Sumo Logic for Linux

> Following are the tags which will be created as part of Linux App install if not already present. 

Reviewer: ```suggestion
Following are the [fields](https://help.sumologic.com/docs/manage/fields/) which will be created as part of Linux App install if not already present. 
```

### Across apps

- **`sumo.datasource`** - has a fixed value of **linux**.

## Collecting logs, metrics & installing Linux app

Here are the steps for collecting logs, metric and installing the app:

### Step 1: Set up Collector

> If you want to use an existing OpenTelemetry Collector then this step can be skipped by selecting the option of using an existing Collector.

Reviewer: ```suggestion
:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the "Use an existing Collector" option.
:::

To create a new Collector:

1. Select the **Add a new Collector** option.
2. Select the platform for which you want to install the Sumo OpenTelemetry Collector.
```

> If you want to create a new Collector please select **Add a new Collector** option.

Reviewer: ```suggestion
```

> Select the platform for which you want to install the Sumo OpenTelemetry Collector.

Reviewer: ```suggestion
```

This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Linux-Collector.png')} alt="Collector" />

### Step 2: Configure integration

In this step we will be configuring the yaml file required for Linux Collection.

The app requires path for system log file, based on the Linux version used:

#### Required Logs for Ubuntu

The following logs, located in your Linux machine's `/var/log` folder, are required for using the Sumo app for Linux with Ubuntu:

> -   `auth.log`
> -   `syslog`
> -   `daemon.log`
> -   `dpkg.log`
> -   `kern.log`
> 
> #### Required Logs for CentOS, Amazon Linux, and Red Hat
> 
> The following logs, located in your Linux machine's `/var/log` folder, are required for using the Sumo app for Linux with CentOS, Amazon Linux, and most Red Hat forks:
> 
> -   audit/audit.log
> -   secure
> -   Messages
> -   yum.log

Reviewer 2: ```suggestion
- `auth.log`
- `syslog`
- `daemon.log`
- `dpkg.log`
- `kern.log`

#### Required Logs for CentOS, Amazon Linux, and Red Hat

The following logs, located in your Linux machine's `/var/log` folder, are required for using the Sumo app for Linux with CentOS, Amazon Linux, and most Red Hat forks:
- audit/audit.log
- secure
- Messages
- yum.log
```

-   By default the collector will be sending process metrics to Sumo Logic. Since the number of processes running can be very large, this may result in significant increase in Data Points per Minute (DPM) If you would like to narrow down the list of processes being monitored, this can be done by adding the following entry under the process section of the downloaded yaml.

```sh
process:
  include:
    names: [ <process name1>, <process name2> ... ]
    match_type: <strict|regexp>
```
:::

### Step 3: Sending logs and metrics to Sumo

Once you have the yaml file downloaded in step 2, follow the below steps :

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Linux instance which needs to be monitored.
2.  restart the collector using
```sh
 sudo systemctl restart otelcol-sumo
```

After successful execution of the above command, Sumo will start receiving the data from your host machine.

> Press "Next". This will install the app to your Sumo Logic Org. The app consists of dashboard and monitors

Reviewer: ```suggestion
Click "**Next**". This will install the app to your Sumo Logic Org. The app consists of dashboard and monitors
```

```

Sample OpenTelemetry metric 

```sql
{"queryId":"A","_source":"linux-otel-metric","process.executable.name":"apache2","_sourceName":"Http Input","process.command":"/usr/sbin/apache2","host":"ip-172-31-90-39.ec2.internal","os.type":"linux","sumo.datasource":"linux","process.executable.path":"/usr/sbin/apache2","process.command_line":"/usr/sbin/apache2_-k_start","process.owner":"www-data","_sourceCategory":"Labs/linux-otel/metric","_contentType":"Carbon2","metric":"process.memory.physical_usage","_collectorId":"000000000C984E1A","_sourceId":"0000000042E512AE","unit":"By","_collector":"Labs - linux-otel","process.pid":"26967","max":42295296,"min":536576,"avg":9061120,"sum":144977920,"latest":8069120,"count":16}
```

## Sample Log Query

Log query from  the panel : Total Event Distribution

```sql
%"sumo.datasource"=linux   
| parse regex "\d+\s+\d+:\d+:\d+\s(?<dest_hostname>\S+)\s(?<process_name>\w*)(?:\[\d+\]|):\s+"
|where dest_hostname matches "{{dest_hostname}}" AND process_name matches "{{process_name}}"
|count as Events by dest_hostname
```

## Sample Metric Query

Metric query from the panel : CPU Utilization Over Time
```sql

> sumo.datasource=linux     host=*  metric=system.cpu.utilization state=(user OR system OR wait OR steal OR softirq OR interrupt OR nice) | sum by host | outlier

Reviewer: ```suggestion
sumo.datasource=linux     host.name=*  metric=system.cpu.utilization state=(user OR system OR wait OR steal OR softirq OR interrupt OR nice) | sum by host | outlier
```

```

## Viewing Linux Dashboards

### Host Metrics - Overview

The Host Metrics - Overview dashboard gives you an at-a-glance view of the key metrics like CPU, memory, disk, network, and TCP connections of all your hosts. You can drill down from this dashboard to the Host Metrics - CPU/Disk/Memory/Network/TCP dashboard by using the honeycombs or line charts in all the panels.

Use this dashboard to:

> -   Identify hosts with high CPU, disk, memory utilization, and identify anomalies over time.

Reviewer 2: ```suggestion
- Identify hosts with high CPU, disk, memory utilization, and identify anomalies over time.
```

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Host-Metrics-Overview.png')} alt="Host Metrics - Overview" />

### Host Metrics - CPU

The Host Metrics - CPU dashboard provides a detailed analysis based on CPU metrics. You can drill down from this dashboard to the Process Metrics - Details dashboard by using the honeycombs or line charts in all the panels.

Use this dashboard to:

> -   Identify hosts and processes with high CPU utilization.
> -   Examine CPU usage by type and identify anomalies over time.

Reviewer 2: ```suggestion
- Identify hosts and processes with high CPU utilization.
- Examine CPU usage by type and identify anomalies over time.
```

-   Monitor abnormal spikes in read/write rates.
-   Compare disk throughput across storage devices of a host.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Host-Metrics-Disk.png')} alt="Host Metrics - Disk" />

### Host Metrics - Memory

The Host Metrics - Memory dashboard provides detailed information on host memory usage, memory distribution, and swap space utilization. You can drill down from this dashboard to the Process Metrics - Details dashboard by using the honeycombs or line charts in all the panels.

Use this dashboard to:

-   Identify hosts with high memory utilization.
-   Examine memory distribution (free, buffered-cache, used, total) for a given host.
-   Monitor abnormal spikes in memory and swap utilization.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Host-Metrics-Memory.png')} alt="Host Metrics - Memory" />

### Host Metrics - Network

The Host Metrics - Network dashboard provides detailed information on host network errors, throughput, and packets sent and received.

Use this dashboard to:

> -   Determine top hosts with network errors and dropped packets.
> -   Monitor abnormal spikes in incoming/outgoing packets and bytes sent and received.
> -   Use dashboard filters to compare throughput across the interface of a host.

Reviewer 2: ```suggestion
- Determine top hosts with network errors and dropped packets.
- Monitor abnormal spikes in incoming/outgoing packets and bytes sent and received.
- Use dashboard filters to compare throughput across the interface of a host.
```

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Host-Metrics-Network.png')} alt="Host Metrics - Network" />

### Host Metrics - TCP

The Host Metrics - TCP dashboard provides detailed information around inbound, outbound, open, and established TCP connections.

Use this dashboard to:

> -   Identify abnormal spikes in inbound, outbound, open, or established connections.

Reviewer 2: ```suggestion
- Identify abnormal spikes in inbound, outbound, open, or established connections.
```

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Linux-OpenTelemetry/Host-Metrics-TCP.png')} alt="Host Metrics - TCP" />

### Process Metrics - Details

> The Process Metrics - Details dashboard gives you a detailed view of key process related metrics such as CPU and memory utilization, disk read/write throughput, and major/minor page faults.

Reviewer 2: ```suggestion
The **Process Metrics - Details** dashboard gives you a detailed view of key process related metrics such as CPU and memory utilization, disk read/write throughput, and major/minor page faults.
```

Use this dashboard to:

> -   Determine the number of open file descriptors in all hosts. If the number of open file descriptors reaches the maximum file descriptor limits, it can cause IOException errors.
> -   Identify anomalies in CPU usage, memory usage, major/minor page faults and reads/writes over time.
> -   Troubleshoot memory leaks using the resident set memory trend chart.

Reviewer 2: ```suggestion
- Determine the number of open file descriptors in all hosts. If the number of open file descriptors reaches the maximum file descriptor limits, it can cause IOException errors.
- Identify anomalies in CPU usage, memory usage, major/minor page faults and reads/writes over time.
- Troubleshoot memory leaks using the resident set memory trend chart.
```
